File: weird_captcha_gym/shared_runtime/server/incubator_graders/firewatch_fold.py
from __future__ import annotations

from typing import Any


MECHANIC_ID = "firewatch_fold"
DIRECTIONS = {
    "UP": (0, 1),
    "RIGHT": (1, 0),
    "DOWN": (0, -1),
    "LEFT": (-1, 0),
}


def _point(value: Any) -> tuple[int, int, int]:
    if not isinstance(value, list) or len(value) != 3:
        raise ValueError("player position is malformed")
    if any(isinstance(item, bool) for item in value):
        raise ValueError("boolean position is invalid")
    return int(value[0]), int(value[1]), int(value[2])


def _tower(truth: dict[str, Any]) -> dict[str, Any]:
    tower = truth.get("initial_tower")
    if not isinstance(tower, dict):
        raise ValueError("missing tower")
    return tower


def _index(
    tower: dict[str, Any],
) -> tuple[
    set[tuple[int, int, int]],
    set[tuple[int, int, int]],
    set[tuple[int, int, int]],
    dict[str, set[tuple[int, int, int]]],
]:
    walls: set[tuple[int, int, int]] = set()
    ladders: set[tuple[int, int, int]] = set()
    fires: dict[str, set[tuple[int, int, int]]] = {}
    faces = tower.get("faces")
    if not isinstance(faces, list) or len(faces) != 2:
        raise ValueError("tower must have two faces")
    for face_index, face in enumerate(faces):
        floors = face.get("floors") if isinstance(face, dict) else None
        if not isinstance(floors, list):
            raise ValueError("tower face floors are malformed")
        for floor_index, floor in enumerate(floors):
            if not isinstance(floor, dict):
                raise ValueError("tower floor is malformed")
            for x in floor.get("walls") or []:
                walls.add((face_index, floor_index, int(x)))
            for x in floor.get("ladders") or []:
                ladders.add((face_index, floor_index, int(x)))
            for fire in floor.get("fires") or []:
                if not isinstance(fire, dict):
                    raise ValueError("fire is malformed")
                fire_id = str(fire.get("id") or "")
                if not fire_id:
                    raise ValueError("fire has no id")
                cells = fire.get("cells")
                if not isinstance(cells, list) or not cells:
                    cells = [fire.get("x")]
                for x in cells:
                    fires.setdefault(fire_id, set()).add(
                        (face_index, floor_index, int(x))
                    )
    fire_cells = {point for points in fires.values() for point in points}
    return walls, ladders, fire_cells, fires


def _target(position: tuple[int, int, int], direction: str, width: int) -> tuple[int, int, int]:
    face, floor, x = position
    dx, dy = DIRECTIONS[direction]
    if direction == "RIGHT" and x == width - 1:
        return 1 - face, floor, 0
    if direction == "LEFT" and x == 0:
        return 1 - face, floor, width - 1
    return face, floor + dy, x + dx


def _snapshot(
    position: tuple[int, int, int],
    facing: str,
    cleared: set[str],
    extinguishers_left: int,
    inspection_face: int,
) -> dict[str, Any]:
    return {
        "player": [position[0], position[1], position[2]],
        "facing": facing,
        "cleared_fires": sorted(cleared),
        "extinguishers_left": int(extinguishers_left),
        "inspection_face": int(inspection_face),
    }


def _mismatch(index: int, field: str, expected: Any, actual: Any) -> dict[str, Any]:
    return {
        "graded": True,
        "passed": False,
        "feedback": f"action {index} has inconsistent {field}: expected {expected!r}, got {actual!r}",
    }


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    if str(payload.get("mechanic_id") or "") != MECHANIC_ID:
        return {"graded": True, "passed": False, "feedback": "mechanic mismatch"}
    if str(ground_truth.get("mechanic_id") or "") != MECHANIC_ID:
        return {"graded": True, "passed": False, "feedback": "ground-truth mechanic mismatch"}
    if str(public_state.get("mechanic_id") or "") != MECHANIC_ID:
        return {"graded": True, "passed": False, "feedback": "public-state mechanic mismatch"}
    challenge_id = str(ground_truth.get("challenge_id") or "")
    task_id = str(ground_truth.get("task_id") or "")
    if not challenge_id or str(payload.get("challenge_id") or "") != challenge_id:
        return {"graded": True, "passed": False, "feedback": "stale challenge"}
    if str(public_state.get("challenge_id") or "") != challenge_id:
        return {"graded": True, "passed": False, "feedback": "public-state challenge mismatch"}
    if not task_id or str(payload.get("task_id") or "") != task_id or str(public_state.get("task_id") or "") != task_id:
        return {"graded": True, "passed": False, "feedback": "task identity mismatch"}
    if public_state.get("tower") != ground_truth.get("initial_tower"):
        return {"graded": True, "passed": False, "feedback": "visible and replay tower geometry differ"}
    if public_state.get("extinguisher_count") != ground_truth.get("extinguisher_count"):
        return {"graded": True, "passed": False, "feedback": "extinguisher stock differs from replay"}

    condition = ground_truth.get("control_condition")
    if condition != public_state.get("control_condition"):
        return {"graded": True, "passed": False, "feedback": "interaction condition differs between public and hidden state"}
    interaction = str((condition or {}).get("interaction") or ground_truth.get("interaction") or "full")
    expected_sources = {
        "MOVE": "keyboard" if interaction == "full" else "control_buttons",
        "EXTINGUISH": "keyboard" if interaction == "full" else "extinguish_button",
        "INSPECT": "eye_button",
    }
    if interaction not in {"full", "simplified"}:
        return {"graded": True, "passed": False, "feedback": "invalid interaction condition"}

    try:
        tower = _tower(ground_truth)
        walls, ladder_cells, fire_cells, fire_positions = _index(tower)
        position = _point(tower.get("start"))
        resident = _point(tower.get("resident"))
        width = int(tower.get("floor_width"))
        floor_count = int(tower.get("floor_count"))
        stock = int(ground_truth.get("extinguisher_count"))
    except (TypeError, ValueError, KeyError) as exc:
        return {"graded": True, "passed": False, "feedback": f"invalid tower contract: {exc}"}

    actions = payload.get("actions")
    if not isinstance(actions, list) or not (1 <= len(actions) <= 900):
        return {"graded": True, "passed": False, "feedback": "action transcript is missing or outside limits"}

    facing = "RIGHT"
    inspection_face = position[0]
    cleared: set[str] = set()
    movement_count = 0
    extinguish_count = 0
    inspection_count = 0

    for index, event in enumerate(actions, start=1):
        if not isinstance(event, dict):
            return {"graded": True, "passed": False, "feedback": f"action {index} is not an object"}
        if event.get("sequence") != index:
            return _mismatch(index, "sequence", index, event.get("sequence"))
        action = str(event.get("action") or "")
        if action not in expected_sources:
            return {"graded": True, "passed": False, "feedback": f"action {index} has invalid type"}
        if event.get("input_source") != expected_sources[action]:
            return {
                "graded": True,
                "passed": False,
                "feedback": f"action {index} used {event.get('input_source')!r}, expected {expected_sources[action]!r}",
            }
        before = _snapshot(position, facing, cleared, stock, inspection_face)
        outcome = ""
        direction = event.get("direction")
        if action == "MOVE":
            movement_count += 1
            if not isinstance(direction, str) or direction not in DIRECTIONS:
                return {"graded": True, "passed": False, "feedback": f"action {index} has invalid movement direction"}
            facing = str(direction)
            target = _target(position, facing, width)
            if target in walls:
                outcome = "blocked_wall"
            elif target in fire_cells and next(
                (
                    fire_id
                    for fire_id, points in fire_positions.items()
                    if target in points
                ),
                None,
            ) not in cleared:
                outcome = "blocked_fire"
            elif direction in {"UP", "DOWN"}:
                ladder_position = position if direction == "UP" else target
                if ladder_position not in ladder_cells or not (0 <= target[1] < floor_count):
                    outcome = "blocked_ladder"
                else:
                    position = target
                    outcome = "climbed"
            else:
                position = target
                outcome = "seam_crossed" if target[0] != before["player"][0] else "walked"
        elif action == "EXTINGUISH":
            if direction != facing:
                return _mismatch(index, "direction", facing, direction)
            target = _target(position, facing, width)
            fire_id = next(
                (
                    item
                    for item, points in fire_positions.items()
                    if target in points
                ),
                None,
            )
            if fire_id is None or fire_id in cleared:
                outcome = "no_fire"
            elif stock <= 0:
                outcome = "stock_empty"
            else:
                extinguish_count += 1
                cleared.add(fire_id)
                stock -= 1
                outcome = "extinguished"
        else:
            inspection_count += 1
            expected_face = 1 - position[0]
            if event.get("face") != expected_face:
                return _mismatch(index, "face", expected_face, event.get("face"))
            inspection_face = expected_face
            outcome = "inspected"

        after = _snapshot(position, facing, cleared, stock, inspection_face)
        expected = {
            "sequence": index,
            "action": action,
            "input_source": expected_sources[action],
            "before": before,
            "after": after,
            "outcome": outcome,
        }
        if action in {"MOVE", "EXTINGUISH"}:
            expected["direction"] = direction
        if action == "INSPECT":
            expected["face"] = inspection_face
        for field, expected_value in expected.items():
            if event.get(field) != expected_value:
                return _mismatch(index, field, expected_value, event.get(field))

    expected_final = _snapshot(position, facing, cleared, stock, inspection_face)
    if payload.get("final_state") != expected_final:
        return {"graded": True, "passed": False, "feedback": "submitted final state does not match replay"}
    try:
        submitted_used = int(payload.get("extinguishers_used"))
    except (TypeError, ValueError):
        return {"graded": True, "passed": False, "feedback": "extinguisher usage is invalid"}
    if submitted_used != extinguish_count or submitted_used != int(ground_truth.get("extinguisher_count")) - stock:
        return {"graded": True, "passed": False, "feedback": "extinguisher usage does not match replay"}
    completed = payload.get("completed") is True
    passed = completed and position == resident and movement_count > 0 and stock >= 0
    return {
        "graded": True,
        "passed": passed,
        "feedback": (
            f"resident {'reached' if position == resident else 'not reached'}; "
            f"face inspections {inspection_count}; fires cleared {len(cleared)}; "
            f"extinguishers left {stock}; actions {len(actions)}"
        ),
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    return {
        "route": ground_truth.get("solution_actions") or [],
        "instruction": "Choose a route whose fire cost fits the stock, walk the ladders, and reach the resident.",
        "answers": [],
    }

File: weird_captcha_gym/shared_runtime/app/mechanics/firewatch_fold.js
(() => {
  "use strict";

  const DELTAS = {
    UP: [0, 1],
    RIGHT: [1, 0],
    DOWN: [0, -1],
    LEFT: [-1, 0],
  };
  const KEYS = {UP: "ArrowUp", RIGHT: "ArrowRight", DOWN: "ArrowDown", LEFT: "ArrowLeft"};
  let helpersCache = null;
  let keyHandler = null;
  let model = null;

  function pointKey(face, floor, x) {
    return `${Number(face)},${Number(floor)},${Number(x)}`;
  }

  function clonePoint(point) {
    return [Number(point[0]), Number(point[1]), Number(point[2])];
  }

  function positionKey(position) {
    return pointKey(position[0], position[1], position[2]);
  }

  function currentFloor(face, floor) {
    return model.state.tower.faces[face]?.floors[floor] || {walls: [], ladders: [], fires: []};
  }

  function hasValue(values, value) {
    return (values || []).some((item) => Number(item) === Number(value));
  }

  function fireAt(position) {
    const floor = currentFloor(position[0], position[1]);
    return (floor.fires || []).find((fire) =>
      Number(fire.x) === Number(position[2])
      || (Array.isArray(fire.cells) && fire.cells.some((x) => Number(x) === Number(position[2])))
    ) || null;
  }

  function fireIsCleared(fire) {
    return Boolean(fire && model.clearedFires.has(String(fire.id)));
  }

  function wallAt(position) {
    return hasValue(currentFloor(position[0], position[1]).walls, position[2]);
  }

  function targetFor(position, direction) {
    const width = Number(model.state.tower.floor_width);
    const [dx, dy] = DELTAS[direction];
    let face = Number(position[0]);
    let floor = Number(position[1]) + dy;
    let x = Number(position[2]) + dx;
    if (direction === "RIGHT" && position[2] === width - 1) {
      face = 1 - face;
      x = 0;
    } else if (direction === "LEFT" && position[2] === 0) {
      face = 1 - face;
      x = width - 1;
    }
    return [face, floor, x];
  }

  function ladderAllows(position, target, direction) {
    const ladderFloor = direction === "UP" ? position[1] : target[1];
    if (ladderFloor < 0 || ladderFloor >= Number(model.state.tower.floor_count) - 1) return false;
    return hasValue(currentFloor(position[0], ladderFloor).ladders, position[2]);
  }

  function snapshot() {
    return {
      player: clonePoint(model.player),
      facing: model.facing,
      cleared_fires: Array.from(model.clearedFires).sort(),
      extinguishers_left: Number(model.extinguishersLeft),
      inspection_face: Number(model.inspectionFace),
    };
  }

  function renderMap() {
    const map = document.getElementById("firewatch-map");
    if (!map) return;
    const tower = model.state.tower;
    map.style.setProperty("--firewatch-cols", String(Number(tower.floor_width || 9)));
    const face = Number(model.inspectionFace);
    const faceData = tower.faces[face];
    const floors = [];
    for (let floorIndex = Number(tower.floor_count) - 1; floorIndex >= 0; floorIndex -= 1) {
      const floor = faceData.floors[floorIndex] || {};
      const cells = [];
      for (let x = 0; x < Number(tower.floor_width); x += 1) {
        const cell = [face, floorIndex, x];
        const key = positionKey(cell);
        const wall = hasValue(floor.walls, x);
        const ladder = hasValue(floor.ladders, x);
        const fire = (floor.fires || []).find((item) =>
          Number(item.x) === x
          || (Array.isArray(item.cells) && item.cells.some((cell) => Number(cell) === x))
        ) || null;
        const cleared = fireIsCleared(fire);
        const player = model.player[0] === face && model.player[1] === floorIndex && model.player[2] === x;
        const resident = tower.resident[0] === face && tower.resident[1] === floorIndex && tower.resident[2] === x;
        const classes = ["firewatch-cell"];
        if (wall) classes.push("is-wall");
        if (ladder && !wall) classes.push("has-ladder");
        if (fire && !cleared && !wall) classes.push("has-fire");
        if (fire && cleared) classes.push("is-cleared");
        if (player) classes.push("has-firefighter");
        if (resident) classes.push("has-resident");
        cells.push(`<div class="${classes.join(" ")}" data-face="${face}" data-floor="${floorIndex}" data-x="${x}">
          <span class="firewatch-cell-number">${x + 1}</span>
          ${wall ? '<span class="firewatch-wall-mark">▦</span>' : ""}
          ${ladder && !wall ? '<span class="firewatch-ladder-mark">╫</span>' : ""}
          ${fire && !cleared && !wall ? '<span class="firewatch-fire-mark">✹</span>' : ""}
          ${fire && cleared ? '<span class="firewatch-ash-mark">✦</span>' : ""}
          ${resident ? '<span class="firewatch-resident-mark">♙</span>' : ""}
          ${player ? '<span class="firewatch-firefighter-mark">✚</span>' : ""}
        </div>`);
      }
      floors.push(`<section class="firewatch-floor" data-floor="${floorIndex}">
        <header><span>F${String(floorIndex + 1).padStart(2, "0")}</span><i>${floorIndex === Number(tower.floor_count) - 1 ? "ROOF" : floorIndex === 0 ? "GROUND" : "SERVICE LEVEL"}</i></header>
        <div class="firewatch-floor-cells">${cells.join("")}</div>
      </section>`);
    }
    map.dataset.inspectionFace = String(face);
    map.innerHTML = floors.join("");
  }

  function renderHud() {
    const position = document.getElementById("firewatch-position");
    const facing = document.getElementById("firewatch-facing");
    const stock = document.getElementById("firewatch-stock");
    const inspected = document.getElementById("firewatch-inspected-face");
    const inspectButton = document.getElementById("firewatch-inspect");
    const cert = document.getElementById("firewatch-certify");
    if (position) position.textContent = `F${String(model.player[1] + 1).padStart(2, "0")} · ${model.state.tower.faces[model.player[0]].label}`;
    if (facing) facing.textContent = `${model.facing} ${({UP: "↑", RIGHT: "→", DOWN: "↓", LEFT: "←"})[model.facing]}`;
    if (stock) {
      stock.textContent = `${model.extinguishersLeft} / ${model.state.extinguisher_count}`;
      stock.dataset.empty = model.extinguishersLeft === 0 ? "true" : "false";
    }
    if (inspected) inspected.textContent = model.state.tower.faces[model.inspectionFace].label;
    const inspectedDuplicate = document.getElementById("firewatch-inspected-face-duplicate");
    if (inspectedDuplicate) inspectedDuplicate.textContent = model.state.tower.faces[model.inspectionFace].label;
    if (inspectButton) inspectButton.textContent = `◉ INSPECT ${model.state.tower.faces[1 - model.player[0]].label}`;
    const reached = model.player[0] === model.state.tower.resident[0] && model.player[1] === model.state.tower.resident[1] && model.player[2] === model.state.tower.resident[2];
    if (cert) cert.classList.toggle("is-ready", reached);
  }

  function renderAll() {
    renderMap();
    renderHud();
  }

  function addEvent(action, inputSource, before, outcome, extra = {}) {
    const event = {
      sequence: model.events.length + 1,
      action,
      input_source: inputSource,
      before,
      after: snapshot(),
      outcome,
      ...extra,
    };
    model.events.push(event);
    renderAll();
    helpersCache.setReadout("ACTION RECORDED", "idle");
  }

  function recordMove(direction, inputSource) {
    if (!model || model.submitting || model.terminal || !DELTAS[direction]) return;
    const before = snapshot();
    model.facing = direction;
    const target = targetFor(model.player, direction);
    let outcome;
    if (wallAt(target)) {
      outcome = "blocked_wall";
    } else if (fireAt(target) && !fireIsCleared(fireAt(target))) {
      outcome = "blocked_fire";
    } else if (direction === "UP" || direction === "DOWN") {
      if (!ladderAllows(model.player, target, direction) || target[1] < 0 || target[1] >= Number(model.state.tower.floor_count)) {
        outcome = "blocked_ladder";
      } else {
        model.player = target;
        outcome = "climbed";
      }
    } else {
      model.player = target;
      outcome = target[0] !== before.player[0] ? "seam_crossed" : "walked";
    }
    addEvent("MOVE", inputSource, before, outcome, {direction});
  }

  function recordExtinguish(inputSource) {
    if (!model || model.submitting || model.terminal) return;
    const before = snapshot();
    const target = targetFor(model.player, model.facing);
    const fire = fireAt(target);
    let outcome;
    if (!fire || fireIsCleared(fire)) {
      outcome = "no_fire";
    } else if (model.extinguishersLeft <= 0) {
      outcome = "stock_empty";
    } else {
      model.clearedFires.add(String(fire.id));
      model.extinguishersLeft -= 1;
      outcome = "extinguished";
    }
    addEvent("EXTINGUISH", inputSource, before, outcome, {direction: model.facing});
  }

  function inspectOtherFace() {
    if (!model || model.submitting || model.terminal) return;
    const before = snapshot();
    model.inspectionFace = 1 - model.player[0];
    addEvent("INSPECT", "eye_button", before, "inspected", {face: model.inspectionFace});
  }

  async function certify() {
    if (!model || model.submitting || model.terminal) return;
    model.submitting = true;
    const button = document.getElementById("firewatch-certify");
    if (button) button.disabled = true;
    helpersCache.setReadout("REPLAYING RESCUE LEDGER…", "idle");
    const resident = model.state.tower.resident;
    const completed = model.player[0] === resident[0] && model.player[1] === resident[1] && model.player[2] === resident[2];
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      actions: model.events,
      final_state: snapshot(),
      extinguishers_used: Number(model.state.extinguisher_count) - Number(model.extinguishersLeft),
      completed,
      certify_input_source: "certify_button",
    };
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify(payload),
      });
      const result = await response.json();
      if (result.passed === true) {
        model.terminal = true;
        document.querySelector(".firewatch-captcha")?.classList.add("is-pass");
        const verdict = document.getElementById("firewatch-verdict");
        if (verdict) verdict.innerHTML = "<b>PASS</b><span>RESIDENT REACHED</span>";
        helpersCache.setReadout("PASS · RESCUE VERIFIED", "passed");
      } else if (result.passed === false) {
        if (result.state) await helpersCache.render(result.state);
        helpersCache.setReadout("FAIL · FRESH TOWER ISSUED", "error");
      } else {
        model.submitting = false;
        if (button) button.disabled = false;
        helpersCache.setReadout("AUDIT UNAVAILABLE", "error");
      }
    } catch (_error) {
      model.submitting = false;
      if (button) button.disabled = false;
      helpersCache.setReadout("AUDIT LINK OFFLINE", "error");
    }
  }

  async function render(state, helpers) {
    helpersCache = helpers || helpersCache;
    if (!helpersCache) throw new Error("firewatch_fold requires runtime helpers");
    if (keyHandler) window.removeEventListener("keydown", keyHandler);
    const tower = state.tower || {};
    const interaction = state.control_condition?.interaction || state.interaction || "full";
    model = {
      state,
      interaction,
      player: clonePoint(tower.start || [0, 0, 0]),
      facing: "RIGHT",
      inspectionFace: Number((tower.start || [0])[0]),
      clearedFires: new Set(),
      extinguishersLeft: Number(state.extinguisher_count || 0),
      events: [],
      submitting: false,
      terminal: false,
    };
    window.firewatchFoldModel = model;
    document.body.dataset.mechanic = "firewatch-fold";
    document.body.dataset.firewatchPalette = String(tower.palette || 0);
    const controls = interaction === "simplified" ? `<div class="firewatch-direction-pad" aria-label="firefighter direction controls">
      <button type="button" data-move="UP" aria-label="move up">↑<small>UP</small></button>
      <button type="button" data-move="LEFT" aria-label="move left">←<small>LEFT</small></button>
      <button type="button" data-move="DOWN" aria-label="move down">↓<small>DOWN</small></button>
      <button type="button" data-move="RIGHT" aria-label="move right">→<small>RIGHT</small></button>
    </div><button class="firewatch-extinguish" id="firewatch-extinguish" type="button"><span>EXTINGUISH AHEAD</span></button>` : `<div class="firewatch-key-guide"><b>ARROWS / WASD</b><span>KEYBOARD MOVEMENT</span><b>SHIFT</b><span>EXTINGUISH AHEAD</span></div>`;
    helpersCache.app.innerHTML = `<section class="firewatch-captcha" data-interaction="${helpersCache.text(interaction)}" data-challenge-id="${helpersCache.text(state.challenge_id)}" tabindex="0">
      <header class="firewatch-header"><div><span class="firewatch-kicker">EMERGENCY ARCHIVE / TOWER ${helpersCache.text(state.challenge_id).toUpperCase()}</span><h1>FIREWATCH <i>FOLD</i></h1><p>${helpersCache.text(state.prompt)}</p></div><div class="firewatch-stamp"><span>RESCUE LEDGER</span><b>OPEN</b></div></header>
      <main class="firewatch-workspace"><section class="firewatch-map-shell"><div class="firewatch-map-top"><div><span class="firewatch-section-label">VISIBLE ELEVATION</span><h2 id="firewatch-inspected-face">${helpersCache.text(tower.faces?.[0]?.label || "NORTH FACE")}</h2></div><div class="firewatch-map-state"><span>PLAYER</span><b id="firewatch-position">F01 · NORTH FACE</b><span>VIEW</span><b id="firewatch-inspected-face-duplicate">—</b></div></div><div class="firewatch-map" id="firewatch-map" aria-label="tower face map"></div><div class="firewatch-map-legend"><span><i class="legend-fire">✹</i> ACTIVE FIRE</span><span><i class="legend-ladder">╫</i> LADDER</span><span><i class="legend-resident">♙</i> RESIDENT</span><span><i class="legend-cross">✚</i> FIREFIGHTER</span></div><div class="firewatch-verdict" id="firewatch-verdict"><b>LIVE</b><span>MAP REPLAY ARMED</span></div></section>
        <aside class="firewatch-console"><div class="firewatch-console-head"><span>FIELD CONSOLE</span><i>SECTOR ${helpersCache.text(state.challenge_id).slice(0, 4).toUpperCase()}</i></div><section class="firewatch-inventory"><label>EXTINGUISHER STOCK</label><div><b id="firewatch-stock">${Number(state.extinguisher_count || 0)} / ${Number(state.extinguisher_count || 0)}</b><span>ONE-USE CANISTERS</span></div></section><section class="firewatch-facing"><label>FACING</label><b id="firewatch-facing">RIGHT →</b></section><button class="firewatch-inspect" id="firewatch-inspect" type="button">◉ INSPECT OTHER FACE</button>${controls}</aside></main>
      <footer class="firewatch-footer"><div><span id="firewatch-manifest-state">ROUTE UNCONFIRMED</span><div class="readout" data-status="idle">READY</div></div><button class="firewatch-certify" id="firewatch-certify" type="button">${helpersCache.text(state.submit_label || "CERTIFY RESCUE")} →</button></footer>
    </section>`;
    const duplicate = document.getElementById("firewatch-inspected-face-duplicate");
    if (duplicate) duplicate.textContent = tower.faces?.[0]?.label || "NORTH FACE";
    document.querySelectorAll("[data-move]").forEach((button) => button.addEventListener("click", () => recordMove(button.dataset.move, "control_buttons")));
    document.getElementById("firewatch-extinguish")?.addEventListener("click", () => recordExtinguish("extinguish_button"));
    document.getElementById("firewatch-inspect")?.addEventListener("click", inspectOtherFace);
    document.getElementById("firewatch-certify")?.addEventListener("click", certify);
    keyHandler = (event) => {
      if (event.repeat || model?.submitting || model?.terminal || model?.interaction !== "full") return;
      const key = String(event.key || "");
      const lower = key.toLowerCase();
      const direction = {ArrowUp: "UP", ArrowRight: "RIGHT", ArrowDown: "DOWN", ArrowLeft: "LEFT", w: "UP", d: "RIGHT", s: "DOWN", a: "LEFT"}[key] || {w: "UP", d: "RIGHT", s: "DOWN", a: "LEFT"}[lower];
      if (direction) {
        event.preventDefault();
        recordMove(direction, "keyboard");
      } else if (key === "Shift") {
        event.preventDefault();
        recordExtinguish("keyboard");
      }
    };
    window.addEventListener("keydown", keyHandler);
    renderAll();
    document.querySelector(".firewatch-captcha")?.focus();
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.firewatch_fold = {
    rootSelector: ".firewatch-captcha",
    render,
  };
})();
